Add first revision of glean ping schema

mdboom opened the pull request:
This was reviewed in https://github.com/mdboom/mobile-telemetry-sdk-schemas/pull/2, but now I am merging it into its "permanent" location.

mdboom commented:
@fbertsch: Great!  Updated to:

- Add the schema as a template
- Uses UUID from existing templates (that's sort of the only relevant one I found, because e.g. histograms is changing considerably in glean, etc.)
- Adds a snakecase template and uses it (since that might be useful elsewhere)
- Updates parts of the test infrastructure to handle [schemaVer versions](https://github.com/snowplow/iglu/wiki/SchemaVer).  It seems that the code treats the versions as strings everywhere, despite the one unnecessary `tonumber` call.  That probably the most controversial part of this PR.  If that's a bridge too far, we can use integer versions for glean versions.  That's a "nice to have" but not "necessary to have".

mdboom commented:
Not sure how to diagnose the test failure.  Where is the `integration.py` that's being run there?

acmiyaguchi commented:
@mdboom It's this [this integration script](https://github.com/mozilla-services/edge-validator#running-integration-tests) that's being run. We're rethinking how this is run on CI, it can safely be ignored.

fbertsch commented:
I think this is as good a place as any to talk about schema. looking over the SchemaVer definition I don't think it's really relevant for us, for a few reasons:

1. Once a ping is validated, it won't ever be checked against the schema again, so the "update a schema, and historical versions may or may not be different" don't apply to us. Additionally, we own the client so we can guarantee that older versions of the ping do or do not match the schema.
2. We really only care about schema incompatibilities, for example when changing the type of a field, so we need a new version of the table.

I've found that there's a lot of overhead when submitting changes that include complicated versioning schemes. My recommendation is definitely a major version (what we have now - bump that and we get a new table version), and if we want to minor version this schema that's fine too. If that's the case we should add a changelog, and include some directions for updating it in the readme (and ci, if possible). That does make sense for the glean ping schema, since it's going to be owned by the data engineering team entirely.

mdboom commented:
Yeah -- given all this, I think it's probably best to back out the schemaver stuff for now, until there's a complete discussion that it's worth the extra complication.  Do you agree @fbertsch?

fbertsch commented:
@mdboom agreed. We'll stick to just a major version for now and we can tackle it in more detail later.

mdboom commented:
Cool.  I've reworked this PR to just use a single integer version.

mdboom commented on schemas/glean/ping/ping.1.schema.json:
```diff
@@ -0,0 +1,386 @@
+      ]
+    }
+  },
+  "description": "Schema for the ping content sent by Mozilla's glean telemetry SDK\n",
```
Should we file a bug to preserve ordering when instantiating the templates?

fbertsch replied:
In general the ordering is fine, I think we need to special case the json schema-metadata to have it first, like `description` and `$schema`. Probably worth filing an issue

fbertsch replied:
It is basically a documentation feature. Seems pretty simple though.

mdboom commented:
> I don't think we want to actually send pings with docType of ping do we? 

In this case, the schema will all be the same for `metrics`, `events` and `"core"`, though.  I guess we need a good generic name?

mreid-moz commented:
We can create a template with a default name, then include the same base schema for each `docType` we want to register.

mdboom commented:
@mreid-moz: Ok.  I put the schema contents in `templates/include/glean`, and then included that from `templates/glean/{baseline|events|metrics}`.  Does that make sense?

fbertsch commented:
@mdboom Mark and I discussed this a bit. For this Q, let's merge one `docType`; `baseline` or `product-summary` is fine. We'll get that data flowing and make a dashboard to achieve this Q's OKRs.

We need to talk more in-depth about what the end goal for the table structures and partitioning format should look like. I'll followup with you and Alessio, and once we understand what those goals are, we can come up with a better configuration story.

mdboom commented:
> @mdboom Mark and I discussed this a bit. For this Q, let's merge one docType; baseline or product-summary is fine. We'll get that data flowing and make a dashboard to achieve this Q's OKRs.

Ok.  I deleted the `metrics` and `events` docTypes and updated the validation examples to use `baseline`.

fbertsch commented:
@whd can we get this deployed?
